x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/literal/IntervalUtils.java

```diff
@@ -0,0 +1,372 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.sql.expression.literal;
+
+import org.elasticsearch.xpack.sql.SqlIllegalArgumentException;
+import org.elasticsearch.xpack.sql.parser.ParsingException;
+import org.elasticsearch.xpack.sql.tree.Location;
+import org.elasticsearch.xpack.sql.type.DataType;
+import org.elasticsearch.xpack.sql.util.StringUtils;
+
+import java.time.Duration;
+import java.time.Period;
+import java.time.temporal.TemporalAmount;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.elasticsearch.xpack.sql.type.DataType.INTERVAL_DAY_TO_HOUR;
+import static org.elasticsearch.xpack.sql.type.DataType.INTERVAL_DAY_TO_MINUTE;
+import static org.elasticsearch.xpack.sql.type.DataType.INTERVAL_DAY_TO_SECOND;
+import static org.elasticsearch.xpack.sql.type.DataType.INTERVAL_HOUR_TO_MINUTE;
+import static org.elasticsearch.xpack.sql.type.DataType.INTERVAL_HOUR_TO_SECOND;
+import static org.elasticsearch.xpack.sql.type.DataType.INTERVAL_MINUTE_TO_SECOND;
+import static org.elasticsearch.xpack.sql.type.DataType.INTERVAL_YEAR_TO_MONTH;
+
+public final class IntervalUtils {
+
+    /**
+     * Time unit definition - used to remember the initial declaration
+     * for exposing it in ODBC.
+     */
+    public enum TimeUnit {
+        YEAR, MONTH, DAY, HOUR, MINUTE, SECOND, MILLISECOND;
+    }
+
+    private IntervalUtils() {}
+
+    public static TemporalAmount of(Location source, long duration, TimeUnit unit) {
+        // Cannot use Period.of since it accepts int so use plus which accepts long
+        // Further more Period and Duration have inconsistent addition methods but plus is there
+        try {
+            switch (unit) {
+                case YEAR:
+                    return Period.ZERO.plusYears(duration);
+                case MONTH:
+                    return Period.ZERO.plusMonths(duration);
+                case DAY:
+                    return Duration.ZERO.plusDays(duration);
+                case HOUR:
+                    return Duration.ZERO.plusHours(duration);
+                case MINUTE:
+                    return Duration.ZERO.plusMinutes(duration);
+                case SECOND:
+                    return Duration.ZERO.plusSeconds(duration);
+                case MILLISECOND:
+                    return Duration.ZERO.plusMillis(duration);
+                default:
+                    throw new ParsingException(source, "Cannot parse duration [{}]", unit);
```

Comment:
That's just a fallback to make the compiler happy - all the enum values are being specified in the case so that branch should never be used unless we expand the enum without adding a case for it.
So currently, there's no value that I can think of that would trigger the exception and thus no tests.